src/OrchardCore/Orchard.Environment.Shell.Abstractions/Builders/ShellContext.cs

```diff
@@ -27,9 +30,57 @@ public class ShellContext : IDisposable
         /// </summary>
         public IServiceScope CreateServiceScope()
         {
+            if (_disposed)
+            {
+                throw new InvalidOperationException("Can't use CreateServiceScope on a disposed context");
+            }
+
+            if (_released)
+            {
+                throw new InvalidOperationException("Can't use CreateServiceScope on a released context");
+            }
+
             return ServiceProvider.CreateScope();
         }
 
+        /// <summary>
+        /// Whether the <see cref="ShellContext"/> instance has been released, for instance when a tenant is changed.
+        /// </summary>
+        public bool Released => _released;
+
+        /// <summary>
+        /// Returns the number of active requests on this tenant.
+        /// </summary>
+        public int ActiveRequests => _refCount;
+
+        public void RequestStarted()
+        {
+            Interlocked.Increment(ref _refCount);
+        }
+
+        public void RequestEnded()
+        {
+            var refCount = Interlocked.Decrement(ref _refCount);
+
+            if (_released && refCount == 0)
+            {
+                Dispose();
+            }
```

Comment:
I first thought that refCount could be incremented while disposing, but it's not possible because here the shellContext is no more in the shellContext Dictionary and is a candidate to be disposed. So, ok.